airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/discover_utils.py

```diff
@@ -0,0 +1,110 @@
+
+from airbyte_cdk.models import (
+    AirbyteStream,
+)
+import json
+from typing import Iterable, Mapping, Any
+from collections import defaultdict 
+
+
+# See https://docs.oracle.com/en/cloud/saas/netsuite/ns-online-help/section_4407755805.html#SuiteAnalytics-Connect-System-Tables
+# for descriptions of the system tables used
+class NetsuiteODBCTableDiscoverer():
+    def __init__(self, cursor):
+        self.cursor = cursor
+        self.data_type_switcher = {
+          "SMALLINT":{"type": ["integer", "null"]},
+          'BIGINT': {"type": ["integer", "null"]},
+          'INTEGER': {"type": ["integer", "null"]},
+          'VARCHAR': {"type": ["string", "null"]},
+          "VARCHAR2":{"type": ["string", "null"]},
+          "WVARCHAR": {"type": ["string", "null"]},
+          "CHAR": {"type": ["string", "null"]},
+          "CLOB": {"type": ["string", "null"]},
+          "TIMESTAMP" : {"type": ["string", "null"], "format": "date-time"},
+          "DOUBLE": {"type": ["number", "null"]},
+        }
+
+    def get_streams(self) -> Iterable[AirbyteStream]:
+      tables = self.get_tables_with_join()
+      primary_keys = self.find_primary_keys_bulk()
+      for key in tables:
+        table_values = tables[key]
+        primary_key_for_table = primary_keys[key] if key in primary_keys else None
+        yield self.get_table_stream_with_join(key, table_values, primary_key_for_table)
+
+    def find_primary_keys_bulk(self):
+      def default_value():
+        return set()
+      self.cursor.execute(f"SELECT pktable_name, pkcolumn_name FROM OA_FKEYS WHERE pktable_name is not NULL and pkcolumn_name IS NOT NULL and fkcolumn_name IS NULL")
+      rows = self.cursor.fetchall()
+      d = defaultdict(default_value)
+      for row in rows:
+        d[row[0]].add(row[1])
+      return d
+      
+    def get_table_stream_with_join(self, table_name, table_columns, primary_key_column):
+      if primary_key_column is None:
+        primary_key_column = []
+      else:
+        primary_key_column = list(primary_key_column)
+
+      properties = {}
+      table_incremental_column = None
+      for column in table_columns:
+        if (self.is_column_incremental_join(column)):
+          table_incremental_column = column['column_name']
+        properties[column['column_name']] = column['column_type']
+      
+      json_schema = {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "type": "object",
+        "properties": properties,
+      }
+
+      if table_incremental_column is not None:
+        return AirbyteStream(name=table_name, json_schema=json_schema, supported_sync_modes=["full_refresh", "incremental"], source_defined_cursor=True, default_cursor_field=[table_incremental_column], primary_key=primary_key_column)
+      else:
+        return AirbyteStream(name=table_name, json_schema=json_schema, supported_sync_modes=["full_refresh"], primary_key=primary_key_column)
+
+    def get_tables_with_join(self):
```

Comment:
I keep tripping on the method name. How about `get_tables_to_columns_mapping`. Adding a type annotation might also help